File: solutions/1043. Partition Array for Maximum Sum/1043.py
# Time complexity: O(nk)
# Space complexity: O(n)
class Solution:
  def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
    n = len(arr)
    maxSum = [0] * (n + 1)

    for start in range(n - 1, -1, -1):
      currMax = 0
      for i in range(start, min(n, start + k)):
        currMax = max(currMax, arr[i])
        maxSum[start] = max(maxSum[start], 
                            maxSum[i + 1] + currMax * (i - start + 1))

    return maxSum[0]        

File: solutions/1043. Partition Array for Maximum Sum/1043.cpp
// Time complexity: O(nk)
// Space complexity: O(n)
class Solution {
 public:
  int maxSumAfterPartitioning(const vector<int>& arr, int k) {
    int n = arr.size();
    vector<int> maxSum(n + 1, 0);

    for (int start = n - 1; start >= 0; --start) {
      int currMax = 0;
      for (int i = start; i < min(n, start + k); ++i) {
        currMax = max(currMax, arr[i]);
        maxSum[start] = max(maxSum[start], 
                            maxSum[i + 1] + currMax * (i - start + 1));
      }
    }

    return maxSum[0];
  }
};